lgc/builder/ImageBuilder.cpp

```diff
@@ -1115,14 +1115,19 @@ Value *ImageBuilder::CreateImageSampleGather(Type *resultTy, unsigned dim, unsig
   // Create the intrinsic.
   Instruction *imageOp = CreateIntrinsic(intrinsicId, overloadTys, args, nullptr, instName);
 
-  // Add a waterfall loop if needed.
-  SmallVector<unsigned, 2> nonUniformArgIndexes;
+  // Add a waterfall loop or insert a readfirstlane for a uniform index if needed.
+  SmallVector<unsigned, 2> descArgIndexes;
   if (flags & ImageFlagNonUniformImage)
-    nonUniformArgIndexes.push_back(imageDescArgIndex);
+    descArgIndexes.push_back(imageDescArgIndex);
   if (flags & ImageFlagNonUniformSampler)
-    nonUniformArgIndexes.push_back(imageDescArgIndex + 1);
-  if (!nonUniformArgIndexes.empty())
-    imageOp = createWaterfallLoop(imageOp, nonUniformArgIndexes);
+    descArgIndexes.push_back(imageDescArgIndex + 1);
+  if (!descArgIndexes.empty())
+    imageOp = createWaterfallLoop(imageOp, descArgIndexes);
+  else {
+    descArgIndexes.push_back(imageDescArgIndex);
+    descArgIndexes.push_back(imageDescArgIndex + 1);
+    handleUniformIndex(imageOp, descArgIndexes);
+  }
```

Comment:
What if one descriptor is nonuniform and the other one isn't?

Reply:
@nhaehnle For the above case, we should not insert a readfirstlane on the descriptor, right?

Reply:
We should insert a readfirstlane in the same way we do it in the case where both are uniform. So the logic should be: do the `handle/enforceUniformIndex` on the descriptors that are uniform, and then if any are non-uniform, apply `createWaterfallLoop` to them together.

This is where it could help slightly if `handle/enforceUniformIndex` does one descriptor at a time instead of an array. You can do:
```
  if (flags & ImageFlagNonUniformImage)
    descArgIndexes.push_back(imageDescArgIndex);
  else
    handleUniformIndex(imageOp, imageDescArgIndex);
  if (flags & ImageFlagNonUniformSampler)
    descArgIndexes.push_back(imageDescArgIndex + 1);
  else
    handleUniformIndex(imageOp, imageDescArgIndex);
  if (!descArgIndexes.empty())
    imageOp = createWaterfallLoop(imageOp, descArgIndexes);
```